File: Assignments/01/Java/CircularList.java
public class CircularList<T> extends LinearStructure<T>{

    // Constructor
    public CircularList() {
        tail = null;
    }

    // Copy Constructor
    CircularList(CircularList<T> other) throws RemoveException {
        this.tail = null;
        if (other.tail != null) {
            Node<T> currPtr = other.tail.next;
            int count = 0;
            do {
                insert(count++, currPtr.element);
                currPtr = currPtr.next;
            }
            while (currPtr != other.tail.next);
        }
    }


    // Clone
    @Override
    public CircularList<T> clone() {
        CircularList<T> clone = new CircularList<>();
        if (this.tail != null) {
            Node<T> currPtr = this.tail.next;
            int count = 0;
            do {
                try {
                    clone.insert(count++, currPtr.element);
                } catch (RemoveException e) {
                    e.printStackTrace();
                }
                currPtr = currPtr.next;
            }
            while (currPtr != this.tail.next);
        }
        return clone;
    }


    // Insert
    @Override
    public void insert(int index, T element) throws RemoveException {
        if (index < 0 || index > size())
            throw new RemoveException("invalid index");
        else {
            Node<T> newNode = new Node<>(element, null);
            if (tail == null) {
                tail = newNode;
                newNode.next = tail;
            }
            else if (index == 0) {
                newNode.next = tail.next;
                tail.next = newNode;
            }
            else if (index == size()) {
                newNode.next = tail.next;
                tail.next = newNode;
                tail = newNode;
            }
            else {
                Node<T> currPtr = tail.next, previousPtr = null;
                for (int count=0;   count<index;    count++, currPtr = currPtr.next)
                    previousPtr = currPtr;
                newNode.next = currPtr;
                previousPtr.next = newNode;
            }
        }
    }


    // Remove
    @Override
    public T remove(int index) throws RemoveException{
        if (index < 0 || index > size() || tail == null)
            throw new RemoveException("empty structure");
        else {
            Node<T> currPtr = tail.next, previousPtr = null;
            T data_item;
            if (tail.next == tail) {
                data_item = tail.element;
                tail = null;
            }
            else if (index == 0) {
                data_item = currPtr.element;
                tail.next = currPtr.next;
            }
            else {
                for (int count=0;   count<index;    count++, currPtr = currPtr.next )
                    previousPtr = currPtr;

                if (index == size()-1) {
                    data_item = tail.element;
                    previousPtr.next = currPtr.next;
                    tail = null;
                    tail = previousPtr;
                }

                else {
                    data_item = currPtr.element;
                    previousPtr.next = currPtr.next;
                }
            }
            return data_item;
        }
    }


    // isEmpty
    @Override
    public boolean isEmpty() {
        return (tail == null);
    }

    @Override
    // Clear
    public void clear() {
        tail = null;
    }

    // Get Leader
    public Node<T> getLeader() {
        return tail;
    }

    // Print Overload
    @Override
    public String toString() {
        StringBuilder printString = new StringBuilder("[");
        if (tail != null) {
            Node<T> currPtr = tail.next;
            do {
                if (currPtr.next == tail.next)
                    printString.append(currPtr.element);
                else
                    printString.append(currPtr.element).append(",");
                currPtr = currPtr.next;
            }
            while (currPtr != tail.next);
        }
        printString.append("]");
        return printString.toString();
    }


    // Private fields and methods
    private int size(){
        int size =0;
        if (tail != null) {
            Node<T> currPtr = tail.next;
            do {
                currPtr = currPtr.next;
                size++;
            }
            while (currPtr != tail.next);
        }
        return size;
    }

    private Node<T> tail;
}

File: Assignments/01/C++/circularList.cpp
//Provide the implementation for the CircularList class in this file.
#include "circularList.h"


//Constructor
template <class T>
CircularList<T>::CircularList() {
    tail = nullptr;
}


//Copy Constructor
template <class T>
CircularList<T>::CircularList(const CircularList<T> &other) {
    tail = nullptr;
    if (other.tail != nullptr) {
        Node<T> *currPtr = other.tail->next;
        int count = 0;
        do {
            insert(count++, currPtr->element);
            currPtr = currPtr->next;
        }
        while (currPtr != other.tail->next);
    }
}


//Overload assignment operator
template <class T>
CircularList<T> &CircularList<T>::operator=(const CircularList<T> &other) {
    if (this != &other) {
        this->clear();
        Node<T> *currPtr = other.tail->next;
        int count = 0;
        do {
            insert(count++, currPtr->element);
            currPtr = currPtr->next;
        }
        while (currPtr != other.tail->next);
    }
    return *this;
}


//Clone
template <class T>
CircularList<T> &CircularList<T>::clone() {
    CircularList<T> *cloned_list = new CircularList<T>(*this);
    return *cloned_list;
}


//Destructor
template <class T>
CircularList<T>::~CircularList() {
    clear();
}


//Insert
template <class T>
void CircularList<T>::insert(int index, T element) {
    if (index < 0 || index > size())
        throw RemoveException("invalid index");
    else {
        Node<T> *newNode = new Node<T>(element);
        if (tail == nullptr) {
            tail = newNode;
            newNode->next = tail;
        }
        else if (index == 0) {
            newNode->next = tail->next;
            tail->next = newNode;
        }
        else if (index == size()) {
            newNode->next = tail->next;
            tail->next = newNode;
            tail = newNode;
        }
        else {
            Node<T> *currPtr = tail->next, *previousPtr = nullptr;
            for (int count=0;   count<index;    count++, currPtr = currPtr->next)
                previousPtr = currPtr;
            newNode->next = currPtr;
            previousPtr->next = newNode;
        }
    }
}


//Remove
template <class T>
T CircularList<T>::remove(int index) {
    if (index < 0 || index > size() || tail == nullptr)
        throw RemoveException("empty structure");
    else {
        Node<T> *currPtr = tail->next, *previousPtr = nullptr, *temp = nullptr, *tailPtr = tail;
        T data_item;
        if (tail->next == tail) {
            data_item = tail->element;
            delete tail;
            tail = nullptr;
        }
        else if (index == 0) {
            data_item = currPtr->element;
            tail->next = currPtr->next;
            delete currPtr;
        }
        else {
            for (int count=0;   count<index;    count++, currPtr = currPtr->next )
                previousPtr = currPtr;

            if (index == size()-1) {
                data_item = tail->element;
                previousPtr->next = currPtr->next;
                tail = nullptr;
                delete currPtr;
                tail = previousPtr;
            }

            else {
                data_item = currPtr->element;
                previousPtr->next = currPtr->next;
                delete currPtr;
            }
        }
        return data_item;
    }
}


//Empty
template <class T>
bool CircularList<T>::isEmpty() {
    return (tail == nullptr);
}


//Clear
template <class T>
void CircularList<T>::clear() {
    if (tail) {
        Node<T> *currPtr = tail->next;
        while (currPtr != tail) {
            Node<T> *previousPtr = currPtr;
            currPtr = currPtr->next;
            delete previousPtr;
        }
        delete tail;
        tail = nullptr;
    }
}



//Head Getter
template <class T>
Node<T> *CircularList<T>::getLeader() {
    return tail;
}


//Size Getter
template <class T>
int CircularList<T>::size() {
    int size =0;
    if (tail) {
        Node<T> *currPtr = tail->next;
        do {
            currPtr = currPtr->next;
            size++;
        }
        while (currPtr != tail->next);
    }
    return size;
}


//Print
template <class T>
ostream &CircularList<T>::print(ostream& os) {
    os << "[";
    if (tail) {
        Node<T> *currPtr = tail->next;
        do {
            if (currPtr->next == tail->next)
                os << currPtr->element;
            else
                os << currPtr->element << ",";
            currPtr = currPtr->next;
        }
        while (currPtr != tail->next);
    }
    os << "]";
    return os;
}




//Extraction Operator Overload
template <class T>
ostream &operator<<(ostream &os, CircularList<T> &obj) {
    obj.print(os);
    return os;
}

